exchanges/bitget/bitget_websocket.go

```diff
@@ -0,0 +1,1212 @@
+			AssetType:  itemDecoder(wsResponse.Arg.InstrumentType),
+			Exchange:   e.Name,
+			StartTime:  time.UnixMilli(ts),
+			CloseTime:  time.UnixMilli(ts).Add(time.Hour * 24),
+			Interval:   "1d",
+			OpenPrice:  open,
+			ClosePrice: closePrice,
+			HighPrice:  high,
+			LowPrice:   low,
+			Volume:     volume,
+		}
+	}
+	e.Websocket.DataHandler <- resp
+	return nil
+}
+
+// TradeDataHandler handles trade data, as functionality is shared between updates and snapshots
+func (e *Exchange) tradeDataHandler(wsResponse *WsResponse) error {
+	var trades []WsTradeResponse
+	pair, err := pairFromStringHelper(wsResponse.Arg.InstrumentID)
+	if err != nil {
+		return err
+	}
+	err = json.Unmarshal(wsResponse.Data, &trades)
+	if err != nil {
+		return err
+	}
+	resp := make([]trade.Data, len(trades))
+	for i := range trades {
+		resp[i] = trade.Data{
+			Timestamp:    trades[i].Timestamp.Time(),
+			CurrencyPair: pair,
+			AssetType:    itemDecoder(wsResponse.Arg.InstrumentType),
+			Exchange:     e.Name,
+			Price:        trades[i].Price.Float64(),
+			Amount:       trades[i].Size.Float64(),
+			Side:         sideDecoder(trades[i].Side),
+			TID:          strconv.FormatInt(trades[i].TradeID, 10),
+		}
+	}
+	e.Websocket.DataHandler <- resp
+	return nil
+}
+
+// OrderbookDataHandler handles orderbook data, as functionality is shared between updates and snapshots
+func (e *Exchange) orderbookDataHandler(wsResponse *WsResponse) error {
+	var ob []WsOrderBookResponse
+	pair, err := pairFromStringHelper(wsResponse.Arg.InstrumentID)
+	if err != nil {
+		return err
+	}
+	err = json.Unmarshal(wsResponse.Data, &ob)
+	if err != nil {
+		return err
+	}
+	if len(ob) == 0 {
+		return errReturnEmpty
+	}
+	bids, err := levelConstructor(ob[0].Bids)
+	if err != nil {
+		return err
+	}
+	asks, err := levelConstructor(ob[0].Asks)
+	if err != nil {
+		return err
+	}
+	if wsResponse.Action[0] == 's' {
+		ob := orderbook.Book{
+			Pair:                   pair,
+			Asset:                  itemDecoder(wsResponse.Arg.InstrumentType),
+			Bids:                   bids,
+			Asks:                   asks,
+			LastUpdated:            wsResponse.Timestamp.Time(),
+			Exchange:               e.Name,
+			ValidateOrderbook:      e.ValidateOrderbook,
+			ChecksumStringRequired: true,
+		}
+		err = e.Websocket.Orderbook.LoadSnapshot(&ob)
+		if err != nil {
+			return err
+		}
+	} else {
+		update := orderbook.Update{
+			Bids:             bids,
+			Asks:             asks,
+			Pair:             pair,
+			UpdateTime:       wsResponse.Timestamp.Time(),
+			Asset:            itemDecoder(wsResponse.Arg.InstrumentType),
+			GenerateChecksum: e.CalculateUpdateOrderbookChecksum,
+			ExpectedChecksum: uint32(ob[0].Checksum), //nolint:gosec // The exchange sends it as ints expecting overflows to be handled as Go does by default
+		}
+		// Sometimes the exchange returns updates with no new asks or bids, just a checksum and timestamp
+		if len(update.Bids) != 0 || len(update.Asks) != 0 {
```

Comment:
Can drop this and Use field AllowEmpty as those will bump the LastUpdateID and it just lets you know that no action happened on the books between the update frequency.